exchanges/gateio/gateio_websocket.go

```diff
@@ -12,720 +15,942 @@ import (
+			pairs, err = g.GetEnabledPairs(asset.CrossMargin)
+		default:
+			pairs, err = g.GetEnabledPairs(asset.Spot)
+		}
+		if err != nil {
+			return nil, err
+		}
+		for j := range pairs {
 			params := make(map[string]interface{})
-			if strings.EqualFold(channels[i], "depth.subscribe") {
-				params["limit"] = 30
-				params["interval"] = "0.1"
-			} else if strings.EqualFold(channels[i], "kline.subscribe") {
-				params["interval"] = 1800
+			switch channelsToSubscribe[i] {
+			case spotOrderbookChannel:
+				params["level"] = 100
+				params["interval"] = kline.HundredMilliseconds
+			case spotCandlesticksChannel:
+				params["interval"] = kline.FiveMin
+			case spotOrderbookUpdateChannel:
+				params["interval"] = kline.ThousandMilliseconds
 			}
-
-			fpair, err := g.FormatExchangeCurrency(enabledCurrencies[j],
-				asset.Spot)
+			fpair, err := g.FormatExchangeCurrency(pairs[j], asset.Spot)
 			if err != nil {
 				return nil, err
 			}
 
 			subscriptions = append(subscriptions, stream.ChannelSubscription{
-				Channel:  channels[i],
+				Channel:  channelsToSubscribe[i],
 				Currency: fpair.Upper(),
 				Params:   params,
-				Asset:    asset.Spot,
 			})
 		}
 	}
 	return subscriptions, nil
 }
 
-// Subscribe sends a websocket message to receive data from the channel
-func (g *Gateio) Subscribe(channelsToSubscribe []stream.ChannelSubscription) error {
-	payloads, err := g.generatePayload(channelsToSubscribe)
+// handleSubscription sends a websocket message to receive data from the channel
+func (g *Gateio) handleSubscription(event string, channelsToSubscribe []stream.ChannelSubscription) error {
+	payloads, err := g.generatePayload(event, channelsToSubscribe)
 	if err != nil {
 		return err
 	}
-
 	var errs error
 	for k := range payloads {
-		resp, err := g.Websocket.Conn.SendMessageReturnResponse(payloads[k].ID, payloads[k])
+		result, err := g.Websocket.Conn.SendMessageReturnResponse(payloads[k].ID, payloads[k])
 		if err != nil {
 			errs = common.AppendError(errs, err)
 			continue
 		}
-		var response WebsocketAuthenticationResponse
-		err = json.Unmarshal(resp, &response)
-		if err != nil {
+		var resp WsEventResponse
+		if err = json.Unmarshal(result, &resp); err != nil {
 			errs = common.AppendError(errs, err)
-			continue
-		}
-		if response.Result.Status != "success" {
-			errs = common.AppendError(errs, fmt.Errorf("%v could not subscribe to %v",
-				g.Name,
-				payloads[k].Method))
-			continue
+		} else {
+			if resp.Result != nil && resp.Result.Status != "success" {
+				errs = common.AppendError(errs, fmt.Errorf("%s websocket connection: timeout waiting for response with and subscription: %v", g.Name, payloads[k].Channel))
+				continue
+			} else if resp.Error != nil && resp.Error.Code != 0 {
+				errs = common.AppendError(errs, fmt.Errorf("error while %s to channel %s error code: %d message: %s", payloads[k].Event, payloads[k].Channel, resp.Error.Code, resp.Error.Message))
+				continue
+			}
```

Comment:
The way this is being handled his been hiding issues.
When running the application with all SPOT pairs enabled, I had no idea why everything wasn't subscribing properly.

Things aren't working properly as is. I added logging to `resp.Error.Message` and got the following:
```
[ERROR] | WEBSOCKET | 23/03/2023 12:56:53 | unknown currency pair ADD_ETH
[ERROR] | WEBSOCKET | 23/03/2023 12:57:33 | unknown currency pair AMPL3L_USDT
[ERROR] | WEBSOCKET | 23/03/2023 12:57:33 | unknown currency pair AMPL3S_USDT
[ERROR] | WEBSOCKET | 23/03/2023 12:57:34 | unknown currency pair AMPL_USDT
```

- You need to be checking for `eventResponse.Error.Message`. 
- Why does `Result.Status != "success"` mean that the websocket timed out?
- Why are currency pairs considered unknown after they've been added via `UpdateTradablePairs`? We shouldn't have invalid pairs

The problem here is you've made it so when there is a subscription error, Websocket connection fails completely and must reconnect, which causes the same issue the next time.

Please look into subscription handling and run the application to verify that things are running appropriately.